### PR TITLE
Make HTTPError instantiable.

### LINKED ISSUE #1377: typeshed annotations do not permit instantiation of urllib.error.HTTPError
Currently in typeshed, [urllib.error.HTTPError is (accurately) annotated](https://github.com/python/typeshed/blob/master/stdlib/3/urllib/error.pyi#L8) as inheriting both `URLError` and `urllib.response.addinfourl`. The `addinfourl` class inherits `addbase`, which [is annotated](https://github.com/python/typeshed/blob/master/stdlib/3/urllib/response.pyi#L6) as inheriting [typing.BinaryIO](https://github.com/python/typing/blob/master/src/typing.py#L2312). This appears to be an attempt to deal with the fact that `addbase` [actually inherits](https://github.com/python/cpython/blob/master/Lib/urllib/response.py#L14) the (private) `tempfile._TemporaryFileWrapper`, which is supposed to wrap a "file-like object", and [proxies all attribute access to the wrapped object](https://github.com/python/cpython/blob/master/Lib/tempfile.py#L473).

This makes some sense: `BinaryIO` is supposed to be the typing "definition" of what "file-like object" means (at least in binary mode). And this works OK for type-checking code that accesses "file-like" methods on an `HTTPError` instance. But it makes it impossible to actually instantiate an `HTTPError` without a spurious mypy error, because the methods of `typing.BinaryIO` (and its parent `typing.IO`) [are all decorated as @abstractmethod](https://github.com/python/typeshed/blob/master/stdlib/3/typing.pyi#L393), and most of them are not overridden by the typeshed annotation of `addbase` or `addinfourl`.

So if you type-check this code (which works fine at runtime):

```
from io import StringIO
from urllib.error import HTTPError

http_error = HTTPError("", 500, "message", dict(), StringIO(''))
```

mypy gives you:

```
error: Cannot instantiate abstract class 'HTTPError' with abstract attributes '__enter__', '__exit__', ... and 'writelines' (16 methods suppressed)
```

I'm filing an issue rather than going straight to pull request because the correct fix here isn't obvious to me, and I'm hoping someone with more typeshed experience can offer a preferred solution. If so, I'm happy to implement it. The obvious solution seems to be repeating all the method definitions from `typing.BinaryIO` and `typing.IO` in the typeshed annotation for `addbase`, without `@abstractmethod`, but I'm hoping there might be an option that avoids this duplication.

### PR DESCRIPTION
**Pull request by carljm**
Fixes #1377.

Demonstration of fix (`../typeshed` is a clone of typeshed with this branch checked out):

```
$ cat httperror.py 
from io import StringIO
from urllib.error import HTTPError

http_error = HTTPError("", 500, "message", dict(), StringIO(''))

$ mypy httperror.py 
httperror.py:4: error: Cannot instantiate abstract class 'HTTPError' with abstract attributes '__enter__', '__exit__', ... and 'writelines' (16 methods suppressed)

$ mypy --custom-typeshed-dir ../typeshed/ httperror.py 
```

### CODE REVIEW

File: stdlib/3/typing.pyi [hunk 1]
```diff
@@ -404,6 +404,41 @@ class BinaryIO(IO[bytes]):
...
+    def truncate(self, size: Optional[int] = ...) -> int: ...
+    def writable(self) -> bool: ...
+    def write(self, s: AnyStr) -> int: ...
+    def writelines(self, lines: Iterable[AnyStr]) -> None: ...
+    def __next__(self) -> AnyStr: ...
+    def __iter__(self) -> Iterator[AnyStr]: ...
+    def __enter__(self) -> BinaryIO: ...
+    def __exit__(self, t: Optional[Type[BaseException]], value: Optional[BaseException],
+                 # TODO: traceback should be TracebackType but that's defined in types
+                 traceback: Optional[Any]) -> bool: ...
```
Comment (JelleZijlstra) [inline stdlib/3/typing.pyi]: TracebackType is already imported, so it should be fine to use.

**Comment by carljm** [comment]
Good point! Just copied that TODO from BinaryIO. Updated both locations to just use TracebackType.

File: stdlib/3/typing.pyi [hunk 2]
```diff
@@ -404,6 +404,41 @@ class BinaryIO(IO[bytes]):
     @abstractmethod
     def __enter__(self) -> BinaryIO: ...
 
+class _ConcreteBinaryIO(BinaryIO):
```
Comment (gvanrossum) [inline stdlib/3/typing.pyi]: I don't think we should do this. Instead we should make some changes in io.pyi and inherit from io.BufferedRandom when we need a concrete subclass of BinaryIO.

Comment (carljm) [inline stdlib/3/typing.pyi]: Ok, I'll see what I can put together that works.

**Comment by carljm** [comment]
So it turns out that just having `addinfourl` inherit `BufferedRandom` instead of `BinaryIO` fixes the immediate issue, without any other changes to `io.pyi`. Is that a workable fix? I looked briefly at the other TODOs in `io.pyi`, but it seemed like some significant refactoring would be needed to address them and avoid lots of "definition incompatible with base class" errors; doesn't seem like that refactoring needs to be conflated with this issue.

**Comment by gvanrossum** [comment]
So if addinfourl() inherits from io.BufferedIO, it's probably not compatible with something that requires a BinaryIO or IO[bytes]. I'm not sure if that's important?

**Comment by carljm** [comment]
Ugh. Yeah, that probably matters.

It also works to subclass `BytesIO` (which inherits `BinaryIO` already) instead of `BinaryIO`. But that gives it some additional methods from `BytesIO` that it probably shouldn't have.

I'll see what I can figure out with the `io.pyi` refactor.

**Comment by gvanrossum** [comment]
Thanks, given that you only encountered it once it's not a high priority to get it fixed, so then I'd rather do it right or not at all (you can keep the `# type: ignore` in the meantime :-).

**Comment by JelleZijlstra** [comment]
@carljm any updates?

**Comment by gvanrossum** [comment]
This should be easier with the changes to io.pyi that just landed.

**Comment by carljm** [comment]
I played with this for a bit but didn't get very far. If some of the io.pyi cleanup has been done for me, I'll take another look (probably next week).

**Comment by gvanrossum** [comment]
Great. Check out the changes that were made in
https://github.com/python/typeshed/pull/1395

**Comment by JelleZijlstra** [comment]
This was just reported again, so perhaps we should try to apply a quick fix to make HTTPError instantiable. Perhaps we can just drop the IO base class?

**Comment by carljm** [comment]
Yeah, sorry... I took a look at this, realized that the io.pyi cleanup had only been done for Py2 and I care about Py3, so would have to figure out how to correctly reapply that cleanup for Py3, and then ran out of time.

**Comment by gvanrossum** [comment]
Ping? We should probably just close this if nobody's going to get to it.

**Comment by carljm** [comment]
I'm not likely to get back to this soon, sorry. It's an unfortunate wart, but it doesn't crop up often (only one instance in our entire codebase AFAIK) and it's easy enough to `# type: ignore`.

**Comment by gvanrossum** [comment]
OK, given the amount of back and forth this has already led to, I'm closing it.

**Pull request closed without merging**